File: src/api/grades-overview.ts
import { RequestFN } from "~/core/request-function";
import { decodeGradesOverview } from "~/decoders/grades-overview";
import { encodePeriod } from "~/encoders/period";
import { type Period, type GradesOverview, type SessionHandle, TabLocation } from "~/models";

/**
 * Get grades overview for a specific period.
 * Including student's grades with averages and the global averages
 */
export const gradesOverview = async (session: SessionHandle, period: Period): Promise<GradesOverview> => {
  const request = new RequestFN(session, "DernieresNotes", {
    _Signature_: { onglet: TabLocation.Grades },

    donnees: {
      Periode: encodePeriod(period)
    }
  });

  const response = await request.send();
  return decodeGradesOverview(response.data.donnees, session);
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/grades-overview.kt
package ink.literate.pawnote.api

import ink.literate.pawnote.core.RequestFN
import ink.literate.pawnote.decoders.decodeGradesOverview
import ink.literate.pawnote.encoders.encodePeriod
import ink.literate.pawnote.models.GradesOverview
import ink.literate.pawnote.models.Period
import ink.literate.pawnote.models.SessionInformation
import ink.literate.pawnote.models.TabLocation
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

suspend fun gradesOverview(sessionInformation: SessionInformation, period: Period): GradesOverview {
  val request =
      RequestFN(
          sessionInformation,
          "DernieresNotes",
          Json.encodeToString(
              buildJsonObject {
                putJsonObject("_Signature_") { put("onglet", TabLocation.Grades.code) }

                putJsonObject("donnees") { put("Periode", encodePeriod(period)) }
              }))

  val response = request.send()
  return decodeGradesOverview(
      Json.parseToJsonElement(response.data).jsonObject["donnees"]!!.jsonObject, sessionInformation)
}
